pkg/piecestore/rpc/client/api/api.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package api
+
+import (
+	"errors"
+	"io"
+	"log"
+
+	"github.com/zeebo/errs"
+
+	"golang.org/x/net/context"
+
+	pb "storj.io/storj/protos/piecestore"
+)
+
+var serverError = errs.Class("serverError")
+
+// PieceMeta -- Struct containing Piece information from PieceMetaRequest
+type PieceMeta struct {
+	Hash       string
+	Size       int64
+	Expiration int64
+}
+
+// PieceMetaRequest -- Request info about a piece by Hash
+func PieceMetaRequest(ctx context.Context, c pb.PieceStoreRoutesClient, hash string) (*PieceMeta, error) {
+
+	reply, err := c.Piece(ctx, &pb.PieceHash{Hash: hash})
+	if err != nil {
+		return nil, err
+	}
+
+	return &PieceMeta{Hash: reply.Hash, Size: reply.Size, Expiration: reply.Expiration}, nil
+}
+
+// StorePieceRequest -- Upload Piece to Server
+func StorePieceRequest(ctx context.Context, c pb.PieceStoreRoutesClient, hash string, data io.Reader, dataOffset int64, length int64, ttl int64, storeOffset int64) error {
+	stream, err := c.Store(ctx)
+
+	buffer := make([]byte, 4096)
+	for {
+		// Read data from read stream into buffer
+		n, err := data.Read(buffer)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+
+		// Write the buffer to the stream we opened earlier
+		if err := stream.Send(&pb.PieceStore{Hash: hash, Size: length, Ttl: ttl, StoreOffset: storeOffset, Content: buffer[:n]}); err != nil {
```

Comment:
we should only send all of the headers the first time. otherwise we're sending the hash and the size and the offset and everything over and over again and again

Reply:
honestly it probably makes sense to send an initial message with no content with all the headers, and then all of the follow up messages are content without headers. might simplify your server-side code too